lib/mauth/client.rb

```diff
@@ -386,6 +491,49 @@ def signature_valid!(object)
         end
       end
 
+      #TODO also is the request body encoded differently? (seems like no but what about if content type is url encoded?)
+      def signature_valid_v2!(object)
+        # We are in an unfortunate situation in which Euresource is percent-encoding parts of paths, but not
+        # all of them.  In particular, Euresource is percent-encoding all special characters save for '/'.
+        # Also, unfortunately, Nginx unencodes URIs before sending them off to served applications, though
+        # other web servers (particularly those we typically use for local testing) do not.  The various forms
+        # of the expected string to sign are meant to cover the main cases.
+        # TODO:  Revisit and simplify this unfortunate situation.
+
+        original_request_uri = object.attributes_for_signing[:request_url]
+        original_query_string = object.attributes_for_signing[:query_string]
+
+        # craft an expected string-to-sign without doing any percent-encoding
+        expected_no_reencoding = object.string_to_sign_v2(time: object.mcc_time, app_uuid: object.signature_app_uuid)
+
+        # do a simple percent reencoding variant of the path
+        expected_for_percent_reencoding = object.string_to_sign_v2(
+          time: object.mcc_time,
+          app_uuid: object.signature_app_uuid,
+          request_url: CGI.escape(original_request_uri.to_s),
+          query_string: CGI.escape(original_query_string.to_s)
+        )
+
+        # do a moderately complex Euresource-style reencoding of the path
+        expected_euresource_style_reencoding = object.string_to_sign_v2(
+          time: object.mcc_time,
+          app_uuid: object.signature_app_uuid,
+          request_url: euresource_escape(original_request_uri.to_s),
+          query_string: euresource_escape(original_query_string.to_s)
+        )
+
+        pubkey = OpenSSL::PKey::RSA.new(retrieve_public_key(object.signature_app_uuid))
+        begin
+          actual = pubkey.public_decrypt(Base64.decode64(object.signature))
+        rescue OpenSSL::PKey::PKeyError
+          raise InauthenticError, "Public key decryption of signature failed!\n#{$!.class}: #{$!.message}"
+        end
+        # TODO: time-invariant comparison instead of #== ?
```

Comment:
Left this todo in here because I'm not really sure what it means. I don't see how `==` is time variant but maybe I'm missing something. Can anyone advise?

Reply:
I don't understand it either.. maybe the intent was thinking of some way to compare without the timestamp.. but it's from a commit 7 years ago by someone no longer here so I don't see any value in keeping.